flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java

```diff
@@ -58,8 +81,29 @@ public void configure(final Configuration config) {
 
     @Override
     public FileSystem create(final URI fsUri) throws IOException {
-        try (TemporaryClassLoaderContext ignored = TemporaryClassLoaderContext.of(loader)) {
-            return new ClassLoaderFixingFileSystem(inner.create(fsUri), loader);
+        final CompletableFuture<FileSystem> fileSystemFuture = new CompletableFuture<>();
+        EXECUTOR.execute(
+                () -> {
+                    try {
+                        synchronized (inner) {
```

Comment:
Why do we need to sync here? The same thread can only have one context loader...